### ISSUE TITLE
Transaction pool may attempt to add transactions between ledger update and pool revalidation

### ISSUE THREAD

**Issue by mrmr1993** [MinaProtocol/mina#7613]
[This assertion](https://github.com/MinaProtocol/mina/blob/cb0ed008f1d39dba838f3dcc5bcf5ca1f0e30362/src/lib/network_pool/indexed_pool.ml#L829) catches this case, and was reported as happening on our nodes in a testnet in #6951, as well as previously in #4255, #4274,  #4282.

My hypothesis is that
* the ledger had updated, including a change to a particular account, but `Indexed_pool.revalidate` had not yet been run
* that account had transactions pending in the pool
* another transaction for that account was gossiped and was being validated

Suggested fix is to
* store the root hash of the ledger in the indexed pool at every revalidation
* before attempting to add a batch of gossiped transactions, check that the ledger hash matches the one that the pool is up-to-date with
  - if it isn't, run `Indexed_pool.wait_for_next_revalidation : unit -> Ledger_hash.t Deferred.t` (not yet implemented)
* before adding *each transaction*, check that the ledger hash hasn't changed
  - the ledger can potentially change out from underneath us because we're sharing a mutable ledger among several threads :weary:
  - if it has, interrupt the current evaluation (since we have modified our copy of the pool, and revalidation may also have modified it)
  - either restart the evaluation (better option) or throw away the transaction pool diff (worse, easier option)

This is a subtle variation on #6946.

**Comment by deepthiskumar** [comment]
I'm not quite sure if it is the change in ledger that could be causing this. From the stacktrace it looks like this is coming from the call to `add_from_gossip_exn` in `handle_transition_frontier_diff` (\"Called from file \\\"src/lib/network_pool/transaction_pool.ml\\\", line 531, characters 30-193\"\n) so why would the ledger change when processing a best tip diff, unless the node went into bootstrap in which case best_tip_ledger would be None (frontier = None) and `add_from_gossip_exn` wouldn't get called. 
Even if the node constructed a frontier, just changing the ledger (without revalidation) should not cause this condition in `dropped_backtrack` to be false: `Account_nonce.equal (Account_nonce.succ last_queued_nonce) nonce`  because we are checking user command's nonce against the last transaction in the pool. So that means it should have completed `revalidate` which could delete last few transactions from a sender and we could hit this assertion failure. For this, the node should have gone from participation to bootstrap, finished boostrapping and  revalidated the pool during `handle_transition_frontier_diff` which seems unlikely? 

I have another theory-`add_from_gossip_exn` gets called when  trying to add back any local non-committed transactions that might have been removed due to the size constraint (returned by`add_from_backtrack`). So when trying to add them back, say there are two transactions from an account, the first transaction wasn't added because of the `has_sufficient_fee` constraint and the second transaction had a higher fee. Adding the second txn without the first one would cause this condition `Account_nonce.equal (Account_nonce.succ last_queued_nonce) nonce` to fail and  the assertion in the else branch as well. So if we are not adding a transaction then we should not try to add any other transaction from the same fee_payer

**Comment by mrmr1993** [comment]
From the file you attached to #6951:
```json
  {
    "insertId": "257dqzm0xw4vd773o",
    "jsonPayload": {
      "message": "Unhandled top-level exception: $exn\nGenerating crash report",
      "source": {
        "module": "Init__Coda_run",
        "location": "File \"src/app/cli/src/init/coda_run.ml\", line 564, characters 2-26"
      },
      "metadata": {
        "peer_id": "12D3KooWNvW4vqWdDTwWZfvJFJP8qGweKEFsNKdeNZqYdHbRzMQb",
        "pid": 11,
        "port": 10071,
        "host": "34.70.33.57",
        "exn": {
          "backtrace": [
            "Raised at file \"src/lib/network_pool/indexed_pool.ml\", line 829, characters 8-75",
            "Called from file \"src/lib/network_pool/transaction_pool.ml\", line 935, characters 28-256",
            "Called from file \"src/lib/network_pool/transaction_pool.ml\", line 1154, characters 18-30",
            "Called from file \"src/lib/network_pool/network_pool_base.ml\", line 93, characters 15-67",
            "Called from file \"src/lib/pipe_lib/strict_pipe.ml\", line 155, characters 28-37",
            "Called from file \"src/deferred0.ml\", line 56, characters 64-69",
            "Called from file \"src/job_queue.ml\" (inlined), line 131, characters 2-5",
            "Called from file \"src/job_queue.ml\", line 171, characters 6-47"
          ],
          "sexp": "Assert_failure src/lib/network_pool/indexed_pool.ml:829:8"
        }
      },
      "timestamp": "2020-12-04 09:26:38.253028Z",
      "level": "Fatal"
    },
    "resource": {
      "type": "k8s_container",
      "labels": {
        "namespace_name": "hard-fork",
        "container_name": "coda",
        "pod_name": "fish-block-producer-56-8466df7d98-j5cb9",
        "location": "us-central1",
        "project_id": "o1labs-192920",
        "cluster_name": "coda-infra-central1"
      }
    },
```
The second line in the backtrace is [here](https://github.com/MinaProtocol/mina/blob/1cab24a51e171134141e7536e9dc3daefbdc751e/src/lib/network_pool/transaction_pool.ml#L935), within the `Transaction_pool.apply` function.

This indicates that something has caused the nonce in the ledger not to match the nonce in the pool; either something is wrong with (how we're updating) the pool, or the ledger has changed (internally via mutation, or by updates to `best_tip_ledger`). Hence my hypothesis above.

**Comment by mrmr1993** [comment]
To unpack the issue you describe a bit:

> Even if the node constructed a frontier, just changing the ledger (without revalidation) should not cause this condition in `dropped_backtrack` to be false: `Account_nonce.equal (Account_nonce.succ last_queued_nonce) nonce` because we are checking user command's nonce against the last transaction in the pool.

Agreed.

> So that means it should have completed `revalidate` which could delete last few transactions from a sender and we could hit this assertion failure. For this, the node should have gone from participation to bootstrap, finished boostrapping and revalidated the pool during `handle_transition_frontier_diff` which seems unlikely?

`handle_transition_frontier_diff` is completely synchronous (it only returns a `Deferred.t` because the final expression is `Deferred.unit`), so it's not possible for any of this to happen during.

> I have another theory-`add_from_gossip_exn` gets called when trying to add back any local non-committed transactions that might have been removed due to the size constraint (returned by`add_from_backtrack`). So when trying to add them back, say there are two transactions from an account, the first transaction wasn't added because of the `has_sufficient_fee` constraint and the second transaction had a higher fee. Adding the second txn without the first one would cause this condition `Account_nonce.equal (Account_nonce.succ last_queued_nonce) nonce` to fail and the assertion in the else branch as well. So if we are not adding a transaction then we should not try to add any other transaction from the same fee_payer

* The `drop_until_below_max_size` that follows an `add_from_backtrack` generates these. This drops commands from an account in decending order of nonce, so the minimum nonce is either unchanged (no commands or only later commands are dropped) or there are no commands for the account (all commands end up being dropped).
* Thus, once an `add_from_backtrack` succeeds for an account without triggering an assertion, all subsequent `add_from_backtrack`s will also succeed without triggering an assertion either
* There are then some `handle_committed` calls, which drop commands from the start of accounts' command lists. A well-formed diff should use these to modify the lists so that the first command's nonce for each account is equal to the account's nonce in the ledger.
* We attempt to re-add the commands from `drop_until_below_max_size`, using `add_from_gossip_exn`. The minimum nonce of each account's list can only have been changed by `add_from_backtrack` or `handle_committed` (unless all commands were dropped, in which case it is none).

The pool being out of sync with the ledger or the diff being malformed are the only ways that this can fail, as far as I can tell. I can add a check to #7639 to rule out the first of these.

**Comment by mrmr1993** [comment]
This is likely to have been fixed by #7759. Closing.
